### PR TITLE
[#2845] WSGI Dispatcher

### PR DESCRIPTION
**Pull request by amercader**
See #2845 for previous discussion

TL;DR:

A top level middleware named `AskAppDispatcherMiddleware`, implemented using [WSGIParty](http://wsgi-party.readthedocs.org/en/latest/). This wraps the two CKAN apps, the Flask one and the Pylons one. One each request it asks both applications if they can handle it, and the apps answer yes or no and whether core or an extension can handle it. The request is then forwarded to the relevant app following this order of precedence:

Flask Extension > Pylons Extension > Flask Core > Pylons Core

### How it works

All the magic happens on `ckan.config.middleware`. 

1. The main WSGI entry point, `make_app` creates the Flask and Pylons stacks separately and passes the two apps to the `AskAppDispatcherMiddleware` constructor. 
2. This stores a reference to the apps and sends them the "invites", which essentially means calling the `join_party` methods of the `CKANFlask` app and the Pylons `PartylineController` to register their `can_handle_request` methods as the ones the dispatcher middleware will call on each request.
3. When a request comes in, `AskAppDispatcherMiddleware` asks both the Pylons and Flask apps if they can handle it (on its `__call__` method). It calls the `can_handle_request` method of each app passing the WSGI environment.
4. Each application performs the check in their own way, eg on Flask is something like:
    ```
          urls = self.url_map.bind_to_environ(environ)
          try:
              endpoint, args = urls.match()
              ...
    ```
    and on Pylons:
    ```
        pylons_mapper = config['routes.map']
        match_route = pylons_mapper.routematch(environ=environ)
        if match_route:
        ...
    ```
5. Each `can_handle_method` returns a tuple in the form `(True, app_name, origin)`. `app_name` can be either `flask_app` or `pylons_app` and `origin` can be `core` or `extension` depending on where this route was defined. (Note that this is not yet implemented for Flask, as we don't know how extensions will handle this).
6. Based on that information, `AskAppDispatcherMiddleware` forwards the request to relevant app, enforcing the order of precedence mentioned above.

### CODE REVIEW

**Comment by amercader** [comment]
@davidread @rossjones would love your views on this

**Comment by davidread** [comment]
Looks good to me. So this will allow us to start replacing controller actions one-by-one, pretty much.

**Comment by amercader** [comment]
@davidread exactly, that's the idea

**Comment by wardi** [comment]
This looks good. It's some extra complexity but it's written in a way that's easy to remove once we drop pylons support.

**Comment by rossjones** [comment]
Yup, removability is probably the biggest concern for me, but looks like you've got it covered. Any idea how much work would be involved in getting it beyond WIP?

**Comment by amercader** [comment]
@rossjones I have to write a few more tests, and maybe move the CKAN Flask app somewhere else. I don't want to add more stuff on this particular PR

**Comment by amercader** [comment]
This is now ready for review

**Comment by davidread** [comment]
Looks great to me, merging.
